float_accessories/lib/can.lisp

```diff
@@ -160,6 +134,75 @@
     })
 })
 
+(defun get-voltage-curve (cell-type)
+    (cond
+        ((= cell-type 0) { ; Linear
+            (list 4.20 4.05 3.90 3.75 3.60 3.45 3.30 3.15 3.00 2.85 2.7)
+        })
+        ((= cell-type 1) { ; P28A
+            (list 4.14 4.09 3.98 3.88 3.77 3.69 3.63 3.55 3.45 3.24 2.9)
+        })
+        ((= cell-type 2) { ; P30B
+           (list 4.14 4.09 3.98 3.88 3.77 3.69 3.63 3.55 3.45 3.24 2.9)
+        })
+        ((= cell-type 3) { ; P42A
+            (list 4.14 4.05 3.91 3.83 3.74 3.65 3.57 3.48 3.38 3 2.8)
+        })
+        ((= cell-type 4) { ; P45B
+            (list 4.14 4.08 4.00 3.91 3.82 3.73 3.64 3.56 3.45 3.23 2.9)
+        })
+        ((= cell-type 5) { ; P50B
+            (list 4.15 4.047 3.96 3.88 3.79 3.70 3.595 3.466 3.29 3.03 2.85)
+        })
+        ((= cell-type 6) { ; DG40
+            (list 4.15 4.02 3.91 3.83 3.75 3.61 3.49 3.35 3.17 2.81 2.7)
+        })
+        ((= cell-type 7) { ; 50S
+            (list 4.15 4.04 3.90 3.82 3.74 3.64 3.52 3.38 3.16 3.0 2.9)
+        })
+        ((= cell-type 8) { ; VTC6
+            (list 4.14 4.00 3.9 3.8 3.7 3.6 3.5 3.4 3.3 3.1 2.8)
+        })
+        (true { ; Any other value we return Linear
+            (list 4.20 4.05 3.90 3.75 3.60 3.45 3.30 3.15 3.00 2.85 2.7)
+        })
+    )
+)
+
+(defunret init-can () {
+    (var can-devices '())
+    (var original-can-id (get-config 'can-id ))
+    (set-config 'can-id -1)
+    (var init-time (systime))
+    (loopwhile (<= (secs-since init-time) 10) {
+        (if (and (>= original-can-id 0) (<= (secs-since init-time) 5)){
+            (setq can-devices (list original-can-id))
+        }{
+            (setq can-devices (can-scan))
+        })
+        (loopforeach can-id can-devices {
+            (setq discover-can-id can-id)
+            (float-cmd can-id (list (assoc float-cmds 'COMMAND_GET_ALLDATA) 3))
+            (yield 500000)
+            (if (>= (get-config 'can-id ) 0) {
+                (if (not-eq (get-config 'can-id ) original-can-id) {
+                    (write-val-eeprom 'can-id (get-config 'can-id ))
+                    (write-val-eeprom 'crc (config-crc cfg-len))
+                })
+                (fetch-series-cells)
+                (float-cmd can-id (list (assoc float-cmds 'COMMAND_HUMIDITY)))
+                (setq soc-type (get-config 'soc-type))
+                (setq cell-type (get-config 'cell-type))
```

Comment:
The problem with moving this here is now you're not updating it on recv-config. Previously the led process would respawn itself. Not the case for can. Make sure you update in recv-config as well.